### PR TITLE
update/improve travis.yml

### PR DESCRIPTION
**Pull request by ljharb**
add node 10 and 11; use `nvm install-latest-npm`; fix install script so it works on older nodes; cache nvm dir

### CODE REVIEW

**Comment by balupton** [comment]
Hey @ljharb would you be able to add a comment for each line, as my understanding of travis and bash are nowhere near yours apparently

File: README.md [hunk 1]
```diff
@@ -30,24 +30,35 @@ Complete configuration for the different [node.js versions](https://github.com/n
...
 matrix:
   fast_finish: true
   allow_failures:
-    - node_js: "0.8"
-    - node_js: "0.10"
     - node_js: "0.12"
+    - node_js: "0.10"
+    - node_js: "0.8"
+    - node_js: "0.6"
 cache:
```
Comment (balupton) [inline README.md:57]: These days, we can just do `cache: npm` or `cache: yarn` right?

File: README.md [hunk 2]
```diff
@@ -30,24 +30,35 @@ Complete configuration for the different [node.js versions](https://github.com/n
...
+  - "0.6"   # end of life
 matrix:
   fast_finish: true
   allow_failures:
-    - node_js: "0.8"
-    - node_js: "0.10"
     - node_js: "0.12"
+    - node_js: "0.10"
+    - node_js: "0.8"
+    - node_js: "0.6"
```
Comment (balupton) [inline README.md:56]: LGTM

File: README.md [hunk 3]
```diff
@@ -30,24 +30,35 @@ Complete configuration for the different [node.js versions](https://github.com/n
 # https://github.com/nodejs/LTS
 sudo: false
 language: node_js
+os:
+ - linux
```
Comment (balupton) [inline README.md:34]: Why is this needed?

File: README.md [hunk 4]
```diff
@@ -30,24 +30,35 @@ Complete configuration for the different [node.js versions](https://github.com/n
...
+before_install:
+  - 'case "${TRAVIS_NODE_VERSION}" in 0.*) export NPM_CONFIG_STRICT_SSL=false ;; esac'
+  - 'nvm install-latest-npm'
+install:
+  - 'if [ "${TRAVIS_NODE_VERSION}" = "0.6" ] || [ "${TRAVIS_NODE_VERSION}" = "0.9" ]; then nvm install --latest-npm 0.8 && npm install && nvm use "${TRAVIS_NODE_VERSION}"; else npm install; fi;'
```
Comment (balupton) [inline README.md:39]: Could you explain what each segment in the `before_install` and `install` does, as my travis/bash cripts are nowhere near as gooda as yours it seems?

File: README.md [hunk 5]
```diff
@@ -30,24 +30,35 @@ Complete configuration for the different [node.js versions](https://github.com/n
...
 node_js:
-  - "0.8"   # end of life
-  - "0.10"  # end of life
-  - "0.12"  # end of life
-  - "4"     # maintenance lts
-  - "6"     # active lts
+  - "11"    # current
+  - "10"    # active lts
```
Comment (balupton) [inline README.md:42]: LGTM

File: README.md [hunk 6]
```diff
@@ -30,24 +30,35 @@ Complete configuration for the different [node.js versions](https://github.com/n
...
   - "8"     # active lts
-  - "9"     # current
+  - "6"     # active lts
+  - "4"     # end of life lts
+  - "0.12"  # end of life
+  - "0.10"  # end of life
+  - "0.8"   # end of life
+  - "0.6"   # end of life
```
Comment (balupton) [inline README.md:49]: LGTM

**Review by balupton — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (ljharb) [reply README.md:34]: it's not - it's implicit - but it leaves a convenient location for adding OS X or Windows if needed in the future (plus, explicit > implicit)

File: README.md [hunk 7]
```diff
@@ -30,24 +30,35 @@ Complete configuration for the different [node.js versions](https://github.com/n
 # https://github.com/nodejs/LTS
 sudo: false
 language: node_js
+os:
+ - linux
+before_install:
+  - 'case "${TRAVIS_NODE_VERSION}" in 0.*) export NPM_CONFIG_STRICT_SSL=false ;; esac'
```
Comment (ljharb) [inline README.md:36]: this avoids SSL errors when connecting to the npm registry in the latest npm versions that 0.x versions of node support

File: README.md [hunk 8]
```diff
@@ -30,24 +30,35 @@ Complete configuration for the different [node.js versions](https://github.com/n
...
+  - 'nvm install-latest-npm'
```
Comment (ljharb) [inline README.md:37]: this installs the latest supported version of npm for the current node version, with logic that is far more accurate than anything else in the ecosystem :-)

File: README.md (hunk 4 above)
Comment (ljharb) [reply README.md:39]: this is because the latest npm version that 0.6 and 0.9 support, doesn't support scoped modules, so it uses node 0.8 to install them and then switches back.

File: README.md (hunk 1 above)
Comment (ljharb) [reply README.md:57]: for `.npm` perhaps, i haven't tried and mildly prefer the explicit form; we'd need the list anyways for the nvm cache dir, unless travis adds support for some kind of `cache: nvm` thing

Comment (balupton) [reply README.md:57]: Good point.

File: README.md (hunk 8 above)
Comment (balupton) [reply README.md:37]: Perhaps this functionality should go into the [`node-upgrade-npm` script](
https://github.com/bevry/awesome-travis/blob/master/scripts/node-upgrade-npm.bash) or become another script — not sure how their functionality differs.

The code inside this snippet should allow easy embedding of the various [awesome-travis scripts](https://github.com/bevry/awesome-travis#scripts).

I can make the necessary alterations upon feedback.

File: README.md (hunk 4 above)
Comment (balupton) [reply README.md:39]: I'm leaning towards this being another [awesome-travis script](https://github.com/bevry/awesome-travis#scripts) as it conflicts with the existing [`node-install`](https://github.com/bevry/awesome-travis/blob/master/scripts/node-install.bash) and [`node-upgrade-npm`](
https://github.com/bevry/awesome-travis/blob/master/scripts/node-upgrade-npm.bash) functionality.

Would be nice for these two parts to become their own scripts for easy embedding/combination with other [awesome-travis scripts](https://github.com/bevry/awesome-travis#scripts).

I can make the necessary alterations upon feedback.

Comment (ljharb) [reply README.md:39]: You can remove this line; it just means any package with scoped deps won’t be able to successfully install on 0.6.

I’m confused what conflicts tho, since it’s using nvm already inside Travis.

File: README.md (hunk 8 above)
Comment (ljharb) [reply README.md:37]: the logic is baked into nvm; it’s not a simple mapping (it’s multi step) and i keep it up to date as npm releases new versions. tbh I’d just replace that entire upgrade npm script with this nvm command, since it’s always available in Travis.

**Comment by balupton** [comment]
Ok merged, will extract the scripts out, so they are more plug-n-play with the rest of the scripts.

**Comment by balupton** [comment]
Ok all done. I replaced `node-upgrade-npm` with your script, and moved the old code to `node-latest-npm`, and added your new install script to `node-npm-install`. And added their appropriate listings to the readme. Also updated the node versions to the latest as of merging.
